plutus-core/plutus-core/src/PlutusCore/ParserCommon.hs

```diff
@@ -0,0 +1,202 @@
+    showErrorComponent (InvalidConstant ty con) =
+        "Invalid constant: " ++ T.unpack con ++ " of type " ++ T.unpack ty
+
+topSourcePos :: SourcePos
+topSourcePos = initialPos "top"
+
+initial :: ParserState
+initial = ParserState M.empty
+
+-- | Return the unique identifier of a name.
+-- If it's not in the current parser state, map the name to a fresh id
+-- and add it to the state. Used in the Name parser.
+intern :: (MonadState ParserState m, PLC.MonadQuote m)
+    => T.Text -> m PLC.Unique
+intern n = do
+    st <- get
+    case M.lookup n (identifiers st) of
+        Just u -> return u
+        Nothing -> do
+            fresh <- PLC.freshUnique
+            let identifiers' = M.insert n fresh $ identifiers st
+            put $ ParserState identifiers'
+            return fresh
+
+type Parser = ParsecT ParseError T.Text (StateT ParserState PLC.Quote)
+instance (Stream s, PLC.MonadQuote m) => PLC.MonadQuote (ParsecT e s m)
+
+parse :: Parser a -> String -> T.Text -> Either (ParseErrorBundle T.Text ParseError) a
+parse p file str = PLC.runQuote $ parseQuoted p file str
+
+parseQuoted :: Parser a -> String -> T.Text -> PLC.Quote (Either (ParseErrorBundle T.Text ParseError) a)
+parseQuoted p file str = flip evalStateT initial $ runParserT p file str
+
+whitespace :: Parser ()
+whitespace = Lex.space space1 (Lex.skipLineComment "--") (Lex.skipBlockCommentNested "{-" "-}")
+
+lexeme :: Parser a -> Parser a
+lexeme = Lex.lexeme whitespace
+
+symbol :: T.Text -> Parser T.Text
+symbol = Lex.symbol whitespace
+
+lparen :: Parser T.Text
+lparen = symbol "("
+rparen :: Parser T.Text
+rparen = symbol ")"
+
+lbracket :: Parser T.Text
+lbracket = symbol "["
+rbracket :: Parser T.Text
+rbracket = symbol "]"
+
+lbrace :: Parser T.Text
+lbrace = symbol "{"
+rbrace :: Parser T.Text
+rbrace = symbol "}"
+
+inParens :: Parser a -> Parser a
+inParens = between lparen rparen
+
+inBrackets :: Parser a -> Parser a
+inBrackets = between lbracket rbracket
+
+inBraces :: Parser a -> Parser a
+inBraces = between lbrace rbrace
+
+isIdentifierChar :: Char -> Bool
+isIdentifierChar c = isAlphaNum c || c == '_' || c == '\''
+
+-- | Create a parser that matches the input word and returns its source position.
+-- This is for attaching source positions to parsed terms/programs.
+-- getSourcePos is not cheap, don't call it on matching of every token.
+reservedWord ::
+    -- | The word to match
+    T.Text -> Parser SourcePos
+reservedWord w = lexeme $ try $ do
+    p <- getSourcePos
+    void $ string w
+    notFollowedBy (satisfy isIdentifierChar)
+    return p
+
+builtinFunction :: (Bounded fun, Enum fun, Pretty fun) => Parser fun
+builtinFunction = lexeme $ choice $ map parseBuiltin [minBound .. maxBound]
+    where parseBuiltin builtin = try $ string (display builtin) >> pure builtin
+
+version :: Parser (PLC.Version SourcePos)
+version = lexeme $ do
+    p <- getSourcePos
+    x <- Lex.decimal
+    void $ char '.'
```

Comment:
I didn't know you could do that!.  I usually write `_ <- char '.'`, but I've always found that really annoying.